### DISCUSSION TITLE
Draft bot evaluation: a learned base score, pick context, and an adaptive deck builder

### DISCUSSION THREAD

**Discussion by pain-l** [phase-rs/phase#8943] [Ideas]
*Work done with Claude Opus 5.*

## Where this comes from

I've been building a Rust draft engine for several months. It performs well, but
it's narrow by design: it leans on per-set 17Lands data and a hand-authored card
taxonomy for each set, so it only works for the handful of sets someone has
prepared.

Phase has the opposite constraint. It has to work on **any** set, on cubes, and
on custom cards, from release day, with no per-set preparation. So this is an
attempt to bring across what survives that constraint, and to be explicit about
what doesn't.

## The problem

`draft_eval::evaluate_draft_card` scores every vanilla N/N for N at exactly
**3.50**, because the body term is `combat_value / mana_value` and that ratio is
`2.5N / N` for every N. Murder, Shock and Unsummon all score exactly **4.00**,
because `is_direct_removal` is one binary OR across Destroy, DealDamage, Bounce,
Counter, Fight, DestroyAll, DamageAll and DiscardCard. Divination scores 2.00,
below a vanilla 1/1.

The whole output range for creatures is about 3.1 to 5.0, while
`bot_ai::eval_score` adds a flat `+4.0` for on-colour. Colour discipline
outweighs the entire spread of card quality.

Measured on The Hobbit (HOB) against real 17Lands win rates: the shipped heuristic's
**top pick in the set is the 151st best card**, and none of its top five cracks
the real top forty.

## What I did

Fit the weights offline against 17Lands games-in-hand win rate and ship only the
coefficients. No runtime data dependency, no per-set authoring, works on release
day for any set including cubes and custom cards.

This is a pattern the repo already uses: `EvalWeightSet::learned()` is learned
from 17Lands replay data via `scripts/train_eval_weights.py`, and
`draft_eval.rs`'s own docs say *"A draft-tuned `DraftWeights::learned()` is
future work."*

- **20 sets**, the most recent with a Premier Draft dump, **5,452 cards, 15.8M games**
- Features from phase's **parsed AST**, not from Oracle text
- Leave-one-set-out throughout. HOB is held out entirely and is *newer* than
  every training set, so it cannot be contaminated
- Labels computed from the 17Lands public S3 dumps

## Results

All on held-out sets. Spearman is rank correlation: it asks whether the model put
the cards in the right *order*, ignoring the actual values. `+1` is perfect, `0`
is no better than random, `-1` is exactly backwards.

| | leave-one-set-out | HOB (held out) |
|---|---|---|
| **fitted** | **+0.430** | **+0.278** |
| shipped heuristic | +0.264 | +0.231 |
| rarity prior only | +0.233 | +0.151 |

**Show it two cards, does it pick the better one?**

| | leave-one-set-out | HOB |
|---|---|---|
| **fitted** | **65.0%** | **59.4%** |
| shipped heuristic | 59.0% | 57.9% |
| rarity prior only | 58.0% | 55.0% |
| coin flip | 50.0% | 50.0% |

**Deal it a 15-card pack, let it take its top pick, look up that card's real win
rate.** A random pick averages 0.5579; a perfect pick averages 0.6270. How much of
that gap does each model capture?

| | leave-one-set-out | HOB |
|---|---|---|
| **fitted** | **54.5%** | **59.9%** |
| shipped heuristic | 33.3% | 17.0% |
| rarity prior only | 32.4% | 23.5% |

That last table is the one that matters for a draft bot: you take one card from a
pack, so being right about the best cards is everything and the ordering of cards
80 through 120 is irrelevant.

The fit also reproduces the defects above independently, without being told
them: `power` and `toughness` come out **negative** while stats *relative to mana
value* come out positive, which is exactly the term `combat / mana_value`
destroys.

Each model's own top five on HOB, and where those really rank out of 184:

```
 fitted #1    real   8     Smaug the Magnificent
 fitted #2    real   6     Beorn the Fierce
 fitted #3    real   1     Fili the Pathfinder
 fitted #4    real  11     Chief Warg's Company
 fitted #5    real  43     Stone-Giant of High Pass

  phase #1    real 151     Desolation of Smaug
  phase #2    real  43     Stone-Giant of High Pass
  phase #3    real 121     Azog, Moria's Ruin
  phase #4    real  88     Sound the Trumpets
  phase #5    real  68     The Great Goblin
```

## Where the bottleneck moves next

At ~60% of pack gap captured, card quality is **no longer the main thing wrong
with the bot**. What remains is that it has no pick context:

- no creature-count target, so it will draft 20 spells and 3 creatures
- a curve term that is a static `match` on cmc with one early/late flag
- a flat on-colour bonus rather than a penalty scaled by how committed the pool is
- no role coverage, so it never takes the second removal spell over the fifth
  creature

Those decide whether a *drafted deck* is good, and none of them are measured
above. They are the larger half of this proposal.

## What a pick looks like at the end

Everything above and below is one expression. The end state:

```
score(card) = base(card)                     learned, 0-100 scale
            + colour(pool, pick_number)      penalty scaled by commitment
            + curve(pool, card)              is this slot still short?
            + composition(pool, card)        creatures vs spells, by shape
            + role(pool, card)               is the pool short on removal?
```

`base` is the learned score above and the only part that needs training. The
rest are small pure functions of the pool, each independently reviewable, each a
no-op until its own phase lands. Nothing here is a rewrite of the picker; it is
one additive term at a time on top of `eval_score`.

## The deck builder

Three things in `suggest.rs` are fixed constants where they should derive from
what was actually drafted. This is the half of the work that needs no model at
all.

**1. Land count.** Currently a fixed ratio:

```rust
const DEFAULT_DECK_SIZE: usize = 40;
const DEFAULT_SPELLS: usize = 23;
```

Always 17 lands, regardless of what was drafted. A two-drop aggro deck and a deck
averaging 4.2 mana get the same manabase. The colour *split* is proportional to
pip counts, which is reasonable; the *count* never adapts.

The count should instead derive from the deck's mean mana value, with credit for
ramp and card selection, and adders for mana sinks, a splash, and a third colour:

```
lands = deck_size × (base_fraction + per_mv_slope × (mean_mv − 3))
        − ramp_and_filtering_credit
        + 1 if mana-sink heavy, + 1 if splashing, + 1 if three colours
        clamped to 16..18
```

Clamping to 16-18 is deliberately conservative. With sensible coefficients that
bound binds below roughly 2.4 mean mana value and above roughly 3.9, so for the
large majority of Limited decks the formula runs free and the clamp only catches
the extremes. It keeps the blast radius small for a first change, and can widen
later if it proves out.

**2. Curve targets don't adapt.** `select_spells_with_curve` fills fixed slots:

```rust
(0,1 → 2), (2 → 6), (3 → 6), (4 → 4), (5 → 3), (6+ → 2)
```

Same targets whether the pool is aggro or control. It also fills slot by slot from
the global score order, so it can take six mediocre two-drops ahead of six
excellent three-drops. Deriving the pool's shape (average mana value, share of
cheap creatures, share of expensive noncreatures) and selecting an ideal curve per
shape fixes both.

**3. There is no creature count target.** Every `Creature` reference in
`suggest.rs` is a test or commander eligibility. Nothing in spell selection knows
what a creature is. A pool that scores well on spells will produce a deck with too
few bodies to function, and nothing detects it.

All three are pure arithmetic over mana values, card types and the drafted pool.
No per-set data, no taxonomy, no training. They work identically on a cube or a
custom set, and they're independently reviewable from the picker work.

## Proposed sequence

Behaviour-neutral until phase 3, so the cheap part gets reviewed before the risky
part lands. The builder phases are independent of the picker phases and could go
first.

| phase | what | behaviour change |
|---|---|---|
| 0 | design docs under `docs/proposals/draft-bot-evaluation/` | none |
| 1 | AST feature extractor + `DraftWeights::learned()` in `phase-ai` | none, nothing calls it |
| 2 | training script + weights artifact, mirroring `retrain-ai-weights` | none |
| 3 | wire into `bot_ai::eval_score` | **yes**, picker |
| 4a | colour penalty scaled by pool commitment | yes, picker |
| 4b | pool shape, composition target, curve | yes, picker |
| 4c | role coverage | yes, picker |
| 5a | adaptive land count in `suggest.rs` | yes, builder |
| 5b | shape-aware curve targets and a creature-count target | yes, builder |

Phase 1 puts the score on a 0-100 scale (win rate 0.44 maps to 0, 0.64 to 100),
which is what unblocks 4a-4c: established context-layer magnitudes transfer
directly instead of being reinvented.

Coefficients for the builder formulas are left unspecified above on purpose. They
would be fitted and validated against phase's own data in the PR rather than
carried over, since anything carried over would want re-validating anyway.

**Not proposed:** signal reading (which lane is open, from what gets passed). It
needs per-set average-last-seen-at data, reintroducing exactly the runtime
dependency this design avoids. This is the main thing that does **not** survive
the any-set constraint.

Also not proposed for now: deck *grading*, scoring a finished deck and iterating
on it. It's a large piece and worth raising only if 5a/5b land well.

## Maintenance, data and reversibility

Worth stating up front, since it's a maintainer's cost rather than mine:

- **No runtime dependency.** The weights are constants in Rust. If 17Lands changes
  its format or disappears, nothing in phase breaks. Retraining is optional, not
  load-bearing.
- **It degrades gracefully.** The numbers above are from a model tested on a set
  newer than everything it trained on, which is the same position it will be in
  for every future set. Retraining when new sets accumulate is an improvement, not
  a repair.
- **Data provenance.** Labels come from 17Lands' public datasets. Only fitted
  coefficients would be committed. No 17Lands data is redistributed, and nothing
  is fetched at build or run time.
- **One-line off switch.** `DraftWeights::default()` stays as the hand-tuned
  fallback, so reverting is changing which constructor is called.
- **Testing.** Aggregate correlation is blind to ordering defects, so behavioural
  tests sit beside it: a strictly better card at the same cost must score higher,
  a rare body must beat the same body at common, unconditional removal must beat
  tempo bounce. Those catch what a metric cannot.

I'm proposing to do the work, not just the design, and I'd rather land phase 0
and one small phase first than open anything large.

## Honest limits

- This is a useful reordering, not a model that understands a format. Per-fold
  spread is wide, +0.219 to +0.538 across 20 folds.
- Games-in-hand win rate is confounded by deck quality and is measurably biased
  against mana value among commons. It is the standard card-quality proxy, not a
  clean measurement of power.
- Format-specific archetype payoffs are not a property of card text, so no
  text-only model learns "this is great in the UB mill deck". That ceiling sits
  above the current heuristic and below a data-fed picker on a set that has data.

## Open questions

1. **Is this wanted, and if so how should it be authorised?** This is protected
   architecture scope, and `AI-CONTRIBUTOR.md` requires either an explicit
   maintainer appointment or a linked issue labelled `accepted` before any PR
   opens, which is why this is a Discussion and not a PR. If you want it, the
   concrete next step is converting this to an issue and labelling it `accepted`,
   or appointing me to it; I'll then open phase 0 (docs only) against that.
   A "no" or "not now" is a fine answer and costs me nothing.
2. **Model form.** The plan ships a dot product with reviewable constants. A
   gradient-boosted model is worth about +0.015 more on held-out data but breaks
   that property. I recommend the dot product; worth a ruling.
3. **Training script location.** `retrain-ai-weights` reserves
   `data/learned-weights.json` for the existing artifact, so this adds a sibling
   with its own `kind:` tag.
4. **Starting point.** Phases 5a and 5b are independent of the model work and need
   no training. If you'd rather see something small land first, the adaptive land
   count is the smallest self-contained piece in the whole proposal.
